text/0000-named-arguments.md

```diff
@@ -0,0 +1,552 @@
+```
+
+### Being realistic
+
+Instead of looking at how code _could_ be written in carefully crafted APIs, we should look at how code is being written _in reality_. Programmers don't always have time to rack their brains over how to create the most beautiful API. They want to get things done.
+
+Named arguments allow iterating quickly without sacrificing readability, because they are dead simple. There's no need to create new types or make up long function names.
+
+This is particularly true for private functions. Since they are not part of the public API, little thought is often given to making them readable. This makes the code harder to maintain as the code base grows. Ideally, both public and private code should be readable and expressive.
+
+## Alternative: Implement structural records instead
+
+Structural records are an active RFC to define anonymous structs. If they are implemented, we'll be able to write:
+
+```rust
+fn free_fall({ z0, v0, g }: { z0: f64, v0: f64, g: f64 });
+free_fall({ z0: 5.0, v0: 0.0, g: 9.81 });
+
+// equivalent code with named arguments:
+fn free_fall(.z0: f64, .v0: f64, .g: f64);
+free_fall(.z0 = 5.0, .v0 = 0.0, .g = 9.81);
+```
+
+However, there are some major differences:
+
+* Structural records as function parameters are like _mandatory_ named arguments
+
+* Converting positional arguments to a structural record is not backwards compatible
+
+* A structural record emulating named arguments is more verbose, since every identifier is repeated in the function declaration
+
+* Converting positional arguments to named arguments is easier than converting them to a structural record
+
+* While named arguments are just syntactic sugar, structural records are a new kind of type, which can appear anywhere, not just in function parameters.
+
+  (This can be seen both as a benefit and as a downside, since it's more powerful, but also more complex and more effort to implement in the compiler.)
+
+I therefore believe that structural records are not a good replacement for named arguments.
+
+# Prior art
+[prior-art]: #prior-art
+
+Named arguments are available in a lot of languages, including _C#, Dart, Kotlin, Python, R, Ruby, Scala, Smalltalk, Swift_ and _Visual Basic_.
+
+Furthermore, in some languages such as JavaScript, named arguments can be emulated with anonymous objects.
+
+### Syntax in other languages
+
+The leading dot is similar to struct initialization syntax in C, e.g. `{ .city = "Hamilton", .prov = "Ontario" }`.
+
+The `=` sign after the argument name is used in several languages that support named arguments, e.g. Python, Kotlin and Scala. Most other languages use a `:` for this instead.
+
+### Rust macros
+
+Some macros in the standard library have a syntax similar to named arguments, for example:
+
+```rust
+println!("The answer is {x}{y}", x = 4, y = 2);
+```
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+* Should it be allowed to specify named arguments in any order?
```

Comment:
By restructuring it you kinda allow this. I think it is not a good idea without a reason.

Reply:
This is particularly confusing if migrating from positional to named calls. If we have
`fn do_it(.action: Action, .client_timeout: Timeout, .server_timeout: Timeout) { ... }`
which was converted to support positional args, then old code which used
`do_it(myAction, Timeout(3), Timeout(5))`
might be migrated to
`do_it(myAction, .server_timeout = Timeout(3), .client_timeout = Timeout(5))`
introducing exactly the kind of bugs named args are supposed to hinder (instead of erroring).

Reply:
There is always a danger of introducing bugs when refactoring code.

I think that IDEs can provide a code action to add argument names automatically. The IDEs will add the argument names in the correct order, so this bug will not happen.

Furthermore, if the client timeout is supposed to be smaller than the server timeout, this bug is really easy to spot thanks to the explicit argument names. Without named arguments, if someone wrote

```rust
do_it(myAction, Timeout(5), Timeout(3))
```

Then it's much harder to find the bug.